fix: parsing buffer edge case at end of buffer

evadnoob opened the pull request:
When at the edge of the buffered reader(the byte buffer used for parsing), handle when there are not enough characters to check for comment conditions.

Adds a new test case for multi-line statements with comments at the edge of th read buffer.

Adds new tracing function to keep tracing in code.

Makes the parse buffer size configurable by a variable, used only for tests.

evadnoob committed:
fix: adds handling for edge case at in the statement parsing

when at the edge of the buffered reader(the byte buffer used for
parsing), handle when there are not enough characters to check for
comment conditions.

evadnoob commented on database/multistmt/parse.go:
```diff
@@ -2,40 +2,26 @@
 // from the multi-statement migration should be parsed and handled.
 type Handler func(migration []byte) error
 
-func splitWithDelimiter(delimiter []byte) func(d []byte, atEOF bool) (int, []byte, error) {
```
removed, not used

evadnoob committed:
added some comments about carry over buffer

samnguyen-outreach commented on database/multistmt/parse.go:
```diff
@@ -47,7 +33,7 @@ func Parse(reader io.Reader, delimiter []byte, maxMigrationSize int, replacement
 	var err error = nil
 	// buf is the bytes read from input reader
-	buf := make([]byte, 1024)
+	buf := make([]byte, ParseBufSize)
```
nit: we don't need to initialize `buf` here if we do it at the beginning of the loop (right?)

evadnoob replied:
agree, though I added the other init/make call to ensure the buffer is clean in each loop iteration, it's sub-optimal performance, I need to do some more testing with the second make call removed.

evadnoob replied:
this has been cleaned up, extra init remove

samnguyen-outreach commented on database/multistmt/parse.go:
```diff
@@ -58,21 +44,71 @@ func Parse(reader io.Reader, delimiter []byte, maxMigrationSize int, replacement
 		n, err := reader.Read(buf)
+		trace("tmp(2): '%s', buf: %s, discard: %v\n", tmp, buf, discard)
+		// tmp is the carry-over buffer,
+		//if the previous loop iteration had two few characters to make comparisions,
```
```suggestion
		//if the previous loop iteration had too few characters to make comparisons,
```

spelling?

samnguyen-outreach commented on database/multistmt/parse.go:
```diff
@@ -58,21 +44,71 @@ func Parse(reader io.Reader, delimiter []byte, maxMigrationSize int, replacement
+		//if the previous loop iteration had two few characters to make comparisions,
+		//tmp will have the characters at the point the loop iteration was abandoned(
+		//break'd out of)
+		if len(tmp) > 0 {
```
Is this condition actually necessary?  If `len(tmp) == 0` the following code would just no-op right?

evadnoob replied:
Also removed this

samnguyen-outreach commented on database/multistmt/parse.go:
```diff
@@ -58,21 +44,71 @@ func Parse(reader io.Reader, delimiter []byte, maxMigrationSize int, replacement
+		//tmp will have the characters at the point the loop iteration was abandoned(
+		//break'd out of)
+		if len(tmp) > 0 {
+			trace("copying '%s' to buf\n", tmp)
```
```suggestion
			trace("prepending '%s' to buf\n", tmp)
```

to be clear what exactly we are doing

samnguyen-outreach commented on database/multistmt/parse.go:
```diff
@@ -58,21 +44,71 @@ func Parse(reader io.Reader, delimiter []byte, maxMigrationSize int, replacement
+			trace("copying '%s' to buf\n", tmp)
+			buf = append(tmp, buf[:n]...)
+			trace("buf: %s\n", buf)
+			// n needs to include the length of tmp since it was copied into n,
+			// but n originally only held the number of chars read from the
+			// reader.Read(buf) call.
+			n = n + len(tmp)
```
instead of maintaining `n` separately, could we just truncate buf to however many bytes were read?  e.g., above just do this:

```
buf = buf[:n]
```

evadnoob replied:
we could but that's another allocation, possibly two more because need to expand the buf again on the next loop iteration, not that it probably matters. But this code path is run many times, unfortunately we don't cache the migration bytes.

samnguyen-outreach commented on database/multistmt/parse.go:
```diff
@@ -2,40 +2,26 @@
-// StartBufSize is the default starting size of the buffer used to scan and parse multi-statement migrations
-var StartBufSize = 4096
+// ParseBufSize is the buffer size for the multi-statement reader
+var ParseBufSize = 1024
```
This seems awfully small, and we could potentially end up using more memory in the end, since we're not reusing the buffers, they will end up as garbage.

evadnoob replied:
that was temporary, the buf was and is re-used again

samnguyen-outreach commented on database/multistmt/parse.go:
```diff
@@ -58,21 +44,71 @@ func Parse(reader io.Reader, delimiter []byte, maxMigrationSize int, replacement
+				// 2 here is the number of look ahead characters that we use.
+				// This tmp buffer is used to copy over bytes from the current loop
+				// iteration if there are not enough characters to lookahead and find a match
+				if i+1 >= len(buf) {
```
Will this always be `i == len(buf) - 1`?  Not asking to change it, just making sure I understand the intention.

If so, it might read simpler to take this condition out of the for loop, and just always ignore the last character of `buf`.  That way we can remove the condition, e.g., 

```golang
if n > 0  {
  tmp = buf[len(buf)-1]
  for i := 0; i < len(buf)-1; i++ {
     ...
}
```

...not sure if this will actually work, feels like there is an edge somewhere I'm missing

samnguyen-outreach commented on database/multistmt/parse.go:
```diff
@@ -108,13 +150,18 @@ func Parse(reader io.Reader, delimiter []byte, maxMigrationSize int, replacement
 			}
+			trace("tmp(1): '%s'\n", tmp)
 		}
+		a = a + n - len(tmp)
```
can you comment to describe what is happening here

samnguyen-outreach commented on database/multistmt/parse.go:
```diff
@@ -58,21 +44,71 @@ func Parse(reader io.Reader, delimiter []byte, maxMigrationSize int, replacement
 	// completed statements, contents of accum will be dumped in here
 	stmts := make([][]byte, 0, 1000)
 
+	tmp := make([]byte, 0, 10)
```
Why do we initialize `tmp`? It's initialized inside the loop already.

```suggestion
	var tmp []byte
```

evadnoob replied:
Yes, good catch, ty

samnguyen-outreach commented on database/multistmt/parse.go:
```diff
@@ -58,21 +44,71 @@ func Parse(reader io.Reader, delimiter []byte, maxMigrationSize int, replacement
+				// This tmp buffer is used to copy over bytes from the current loop
+				// iteration if there are not enough characters to lookahead and find a match
+				if i+1 >= len(buf) {
+					tmp = make([]byte, n-i)
```
related to my question above, isn't `tmp` always going to be size of 1? (Since we only need to read ahead 1 additional character)

evadnoob replied:
Yes, but I wanted to write the tmp buffer in anticipation of additional look-ahead comparisons and not limit to a single character.

samnguyen-outreach commented on database/multistmt/parse.go:
```diff
@@ -58,21 +44,71 @@ func Parse(reader io.Reader, delimiter []byte, maxMigrationSize int, replacement
+				// This tmp buffer is used to copy over bytes from the current loop
+				// iteration if there are not enough characters to lookahead and find a match
+				if i+1 >= len(buf) {
+					tmp = make([]byte, n-i)
```
related to my question above, isn't `tmp` always going to be size of 1? (Since we only need to read ahead 1 additional character).

If so, could we do this to make it simpler to read?

```suggestion
					tmp = make([]byte, 1)
```

evadnoob replied:
Only if we never do comparisons looking for two additional characters, I'd like to keep the flexibility

samnguyen-outreach commented on database/multistmt/parse.go:
```diff
@@ -98,6 +137,9 @@ func Parse(reader io.Reader, delimiter []byte, maxMigrationSize int, replacement
 							s1 := strings.ReplaceAll(string(c1), "<SCHEMA_NAME>", replacementStatement)
 							c1 = []byte(s1)
 						}
+						// in the future this could be the place to run statements
+						//instead of keeping them as an array(
+						//the array subverts the streaming intention of this reader)
```
Does this mean we don't actually stream the statements? Do they all end up stored in memory?

evadnoob replied:
Yes, that's why I left that comment, we stream the bytes, then accumulate them in an array and then run them. I think it's a trivial change to remove the array and execute the statements during streaming.

samnguyen-outreach reviewed (approved):
Approving, as I don't actually have any blocking comments

evadnoob committed:
Update database/multistmt/parse.go

Co-authored-by: Sam Nguyen <102551395+samnguyen-outreach@users.noreply.github.com>

evadnoob committed:
take some comments from PR review

- fixup comment style
- rename a -> counter
- remove conditional for tmp, it is redundant
- change to execute statements as soon as a fully formed statement is
  available and remove the stmt array

evadnoob committed:
fix byte array, for stmt exec
